overlord/snapstate/aliases.go

```diff
@@ -0,0 +1,338 @@
+	toggleAliases.Set("snap-setup", &snapsup)
+
+	setupAliases := st.NewTask("setup-aliases", fmt.Sprintf(i18n.G("Setup snap %q aliases"), snapsup.Name()))
+	setupAliases.Set("snap-setup", &snapsup)
+	setupAliases.WaitFor(toggleAliases)
+
+	return state.NewTaskSet(toggleAliases, setupAliases), nil
+}
+
+func (m *SnapManager) doToggleAliases(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+	snapsup, snapst, err := snapSetupAndState(t)
+	if err != nil {
+		return err
+	}
+	snapName := snapsup.Name()
+	curInfo, err := snapst.CurrentInfo()
+	if err != nil {
+		return err
+	}
+
+	aliasStates, err := getAliases(st)
+	if err != nil {
+		return err
+	}
+	if aliasStates == nil {
+		aliasStates = make(map[string]*AliasState)
+	}
+	changes := make(map[string]string)
+	for _, alias := range snapsup.Aliases {
+		if curInfo.Aliases[alias] == nil {
+			return fmt.Errorf("cannot enable alias %q for %q, no such alias", alias, snapName)
+		}
+		aliasState := aliasStates.get(alias)
+		if aliasState.Enabled == "" {
+			if err := checkAliasConflict(st, snapName, alias); err != nil {
+				return err
+			}
+			aliasState.Enabled = snapName
+			// TODO: remove from Disabled if it's there etc
+			aliasStates.set(alias, &aliasState)
+			changes[alias] = ">enabled"
+		} else if aliasState.Enabled == snapName {
+			// nothing to do
+		} else {
+			return fmt.Errorf("cannot enable alias %q for %q, already enabled for %q", alias, snapName, aliasState.Enabled)
+		}
+	}
+	if len(changes) != 0 {
+		t.Set("changes", changes)
+		st.Set("aliases", aliasStates)
+	}
+	return nil
+}
+
+func checkAliasConflict(st *state.State, snapName, alias string) error {
+	var snapNames map[string]*json.RawMessage
+	err := st.Get("snaps", &snapNames)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	for name := range snapNames {
+		if name == alias || strings.HasPrefix(alias, name+".") {
+			return fmt.Errorf("cannot enable alias %q for %q, it conflicts with the command namespace of installed snap %q", alias, snapName, name)
+		}
+	}
+	return nil
+}
+
+func (m *SnapManager) undoToggleAliases(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+	var changes map[string]string
+	err := t.Get("changes", &changes)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	if len(changes) == 0 {
+		// nothing to undo
+		return nil
+	}
+	snapsup, _, err := snapSetupAndState(t)
+	if err != nil {
+		return err
+	}
+	snapName := snapsup.Name()
+	aliasStates, err := getAliases(st)
+	if err != nil {
+		return err
+	}
+	if aliasStates == nil {
+		aliasStates = make(map[string]*AliasState)
+	}
+	for alias, change := range changes {
+		aliasState := aliasStates.get(alias)
+		switch change {
+		case ">enabled":
+			if aliasState.Enabled == snapName {
+				aliasState.Enabled = ""
+				aliasStates.set(alias, &aliasState)
+			}
+		case "enabled>":
```

Comment:
The logic around these strings seems pretty obscure, and "toggle" itself is a bit unclear. Have you considered breaking down into "alias" and "unalias" tasks? The code might end up more straightforward, even if there's some additional duplication. We did that with connect/disconnect, I think.